### ISSUE TITLE
Proposal: replace Variable.volatile with global switch

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#3627]
Variable.volatile forces outputs to not require gradients if any of the inputs are marked volatile. This works OK in the forward pass, but we're forced to [change the meaning](https://github.com/pytorch/pytorch/blob/b06c59e543aa26586087c19fb7b713f8872105bb/torch/csrc/autograd/functions/accumulate_grad.cpp#L38) in the backwards. Gradients are sometimes volatile and sometimes not, which is awkward if you add them back to parameters, such as in optimizers.

We should replace `volatile` with a context manager in Python. (Chainer already did this with [`no_backprop_mode()`](http://docs.chainer.org/en/stable/reference/core/generated/chainer.no_backprop_mode.html#chainer.no_backprop_mode)).

At the C++ level, we should replace it with a thread-local global switch.

This will simplify the logic in the backwards: by default `backwards()` will set "no-backprop mode", unless `create_graph` is True.

**Comment by colesbury** [comment]
cc @jekbradbury

**Comment by jekbradbury** [comment]
I'm not sure this is compatible with the proposed unification of Variables and Tensors, in that we'd want the future behavior of a plain tensor to be the same as the current behavior of a `volatile=True` Variable?
The global switch has worked well for Chainer but they continue to make a strong distinction between `cupy.array` and `chainer.Variable`

**Comment by colesbury** [comment]
If we keep `volatile`, the default would be `requires_grad=False`, `volatile=False` (the same as when you construct a `Variable`)

**Comment by jekbradbury** [comment]
Hmm ok, that works too; we’ll have to deprecate the (admittedly strange) concept of using a flag on the data tensor/variable to represent train/test sets (overly magical but I currently rely on it in torchtext). I’m not a huge fan of global state but we have some already in the device contexts. So it looks like this is the right thing to do.

**Comment by ezyang** [comment]
Paging @apaszke; I distinctly remember chatting with him about this when we were writing the PyTorch paper, and when I mentioned to him that Chainer had made it a global flag, Adam seemed pretty convinced that having it as local flags was still a better idea. He managed to convince me at the time but I have now forgotten how I was convinced.

**Comment by apaszke** [comment]
I think volatile works better than context managers and thread local state when you have a clear distinction between Tensor and Variable worlds. I think it will create a wide range of problems once we unify them, so it's worth considering dumping the flag 😕 I just really don't like magic thread-local things.
